File: No17_LetterCombinationsOfAPhoneNumber/LetterCombinationsOfAPhoneNumber.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        letters = [
            "",     # 0
            "",     # 1
            "abc",  # 2
            "def",  # 3
            "ghi",  # 4
            "jkl",  # 5
            "mno",  # 6
            "pqrs", # 7
            "tuv",  # 8
            "wxyz", # 9
        ]

        result: List[str] = []
        idx = [0] * len(digits)

        while True:
            buf = []
            for i, v in enumerate(idx):
                m = letters[ord(digits[i]) - ord('0')]
                buf.append(m[v])
            result.append("".join(buf))

            i = len(idx) - 1
            while i >= 0:
                m = letters[ord(digits[i]) - ord('0')]
                idx[i] += 1
                if idx[i] < len(m):
                    break
                idx[i] = 0
                i -= 1

            if i < 0:
                break

        return result

File: No17_LetterCombinationsOfAPhoneNumber/LetterCombinationsOfAPhoneNumber.go
package main

func letterCombinations(digits string) []string {
    if len(digits) == 0 {
        return []string{}
    }
    
    letters := []string{
        " ",     // 0
        "",      // 1
		"abc",   // 2
		"def",   // 3
		"ghi",   // 4
		"jkl",   // 5
		"mno",   // 6
		"pqrs",  // 7
		"tuv",   // 8
		"wxyz",  // 9
	}

    result := []string{}
    idx := make([]int, len(digits))
    for {
		buf := make([]byte, len(digits))
		for i, v := range idx {
			m := letters[int(digits[i]-'0')]
			buf[i] = m[v]
		}
		result = append(result, string(buf))

        i := len(digits) - 1
        for i >= 0 {
            idx[i]++
            if idx[i] < len(letters[int(digits[i]-'0')]) {
                break
            }
            idx[i] = 0
            i--
        }

        if i < 0 {
            break
        }
    }
    return result
}